src/compiler.ts

```diff
@@ -2999,195 +3000,325 @@ export class Compiler extends DiagnosticEmitter {
-              let scopedLocals = flow.scopedLocals;
-              if (!scopedLocals) flow.scopedLocals = scopedLocals = new Map();
-              else if (scopedLocals.has(name)) {
-                let existing = assert(scopedLocals.get(name));
-                this.errorRelated(
-                  DiagnosticCode.Duplicate_identifier_0,
-                  declaration.name.range,
-                  existing.declaration.name.range,
-                  name
+
+            let idents = pattern.elements;
+            let numIdents = idents.length;
+            for (let j: i32 = 0; j < numIdents; ++j) {
+              let nameNode = idents[j];
+              let isConst = declaration.is(CommonFlags.CONST);
+              if (nameNode.kind == NodeKind.IDENTIFIER) {
+                let name = (<IdentifierExpression>nameNode).text;
+                // add local
+                let local: Local;
+                let type = indexedGet.signature.returnType;
+                if (
+                  declaration.isAny(CommonFlags.LET | CommonFlags.CONST) ||
+                  flow.isInline
+                ) { // here: not top-level
+                  let existingLocal = flow.getScopedLocal(name);
+                  if (existingLocal) {
+                    if (!existingLocal.declaration.range.source.isNative) {
+                      this.errorRelated(
+                        DiagnosticCode.Duplicate_identifier_0,
+                        nameNode.range,
+                        existingLocal.declaration.name.range,
+                        name
+                      );
+                    } else { // scoped locals are shared temps that don't track declarations
+                      this.error(
+                        DiagnosticCode.Duplicate_identifier_0,
+                        nameNode.range, name
+                      );
+                    }
+                    local = existingLocal;
+                  } else {
+                    local = flow.addScopedLocal(name, type);
+                  }
+                  if (isConst) flow.setLocalFlag(local.index, LocalFlags.CONSTANT | LocalFlags.INITIALIZED);
+                } else {
+                  let existing = flow.lookupLocal(name);
+                  if (existing) {
+                    this.errorRelated(
+                      DiagnosticCode.Duplicate_identifier_0,
+                      nameNode.range,
+                      existing.declaration.name.range,
+                      name
+                    );
+                    continue;
+                  }
+                  local = flow.parentFunction.addLocal(type, name, declaration);
+                  if (isConst) flow.setLocalFlag(local.index, LocalFlags.CONSTANT | LocalFlags.INITIALIZED);
+                }
+
+                let callExpr = this.compileCallDirect(indexedGet, [
+                  Node.createIntegerLiteralExpression(i64_new(j), nameNode.range)
```

Comment:
Is there a better way to do this?

Reply:
Perhaps `makeCallDirect` to circumvent the AST and provide a Binaryen IR node directly :)